lib/src/trailer.rs

```diff
@@ -0,0 +1,264 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Parsing trailers from commit messages.
+
+use itertools::Itertools as _;
+
+/// A key-value pair representing a trailer in a commit message, of the
+/// form `Key: Value`.
+#[derive(Debug, PartialEq, Clone)]
+pub struct Trailer {
+    /// trailer key
+    pub key: String,
+    /// normalized trailer value
+    pub value: String,
+}
+
+/// Parse the trailers from a commit message; these are simple key-value
+/// pairs, separated by a colon, describing extra information in a commit
+/// message; an example is the following:
+///
+/// ```text
+/// chore: update itertools to version 0.14.0
+///
+/// Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
+/// tempor incididunt ut labore et dolore magna aliqua.
+///
+/// Co-authored-by: Alice <alice@example.com>
+/// Co-authored-by: Bob <bob@example.com>
+/// Reviewed-by: Charlie <charlie@example.com>
+/// Change-Id: I1234567890abcdef1234567890abcdef12345678
+/// ```
+///
+/// In this case, there are four trailers: two `Co-authored-by` lines, one
+/// `Reviewed-by` line, and one `Change-Id` line.
+pub fn parse_description_trailers(body: &str) -> Vec<Trailer> {
+    let (trailer, blank) = parse_trailers_and_blank(body);
+    if blank {
+        trailer
+    } else {
+        // no blank found, this means there was a single paragraph, so whatever
+        // was found can't come from the trailer
+        vec![]
+    }
+}
+
+/// Parse the trailers from a trailer paragraph. This function behaves like
+/// `parse_description_trailer`, except that it doesn't expect the body to
+/// contain several paragraphs.
+pub fn parse_trailers(body: &str) -> Vec<Trailer> {
+    let (trailer, _) = parse_trailers_and_blank(body);
+    trailer
+}
+
+fn parse_trailers_and_blank(body: &str) -> (Vec<Trailer>, bool) {
+    // a trailer always comes at the end of a message; we can split the message
+    // by newline, but we need to immediately reverse the order of the lines
+    // to ensure we parse the trailer in an unambiguous manner; this avoids cases
+    // where a colon in the body of the message is mistaken for a trailer
+    let lines = body.trim_end().lines().rev();
+    let trailer_re =
+        regex::Regex::new(r"^([a-zA-Z0-9-]+) *: +(.+) *$").expect("trailer regex should be valid");
```

Comment:
nit: might be simpler to use `.split_once()`, `.trim_ascii()`, `.is_ascii_alphanumeric()`, etc., but I don't have strong feeling against regex use.

Reply:
The regex checks that the key is only made of letters, digits, and dashes. It seemed easier to do this with a regex. If that's okay with you, I'll keep it implemented with a regex.